btf/types.go

```diff
@@ -841,62 +847,112 @@ func inflateRawTypes(rawTypes []rawType, rawStrings *stringTable, base *Spec) ([
 		return members, nil
 	}
 
+	var (
+		header    btfType
+		bInt      btfInt
+		bArr      btfArray
+		bMembers  []btfMember
+		bEnums    []btfEnum
+		bParams   []btfParam
+		bVariable btfVariable
+		bSecInfos []btfVarSecinfo
+		bDeclTag  btfDeclTag
+		bEnums64  []btfEnum64
+	)
+
 	var declTags []*declTag
-	for _, raw := range rawTypes {
+	for {
 		var (
 			id  = firstTypeID + TypeID(len(types))
 			typ Type
 		)
 
+		if err := binary.Read(r, bo, &header); err == io.EOF {
+			break
+		} else if err != nil {
+			return nil, fmt.Errorf("can't read type info for id %v: %v", id, err)
+		}
+
 		if id < firstTypeID {
 			return nil, fmt.Errorf("no more type IDs")
 		}
 
-		name, err := rawStrings.Lookup(raw.NameOff)
+		name, err := rawStrings.Lookup(header.NameOff)
 		if err != nil {
 			return nil, fmt.Errorf("get name for type id %d: %w", id, err)
 		}
 
-		switch raw.Kind() {
+		switch header.Kind() {
 		case kindInt:
-			size := raw.Size()
-			bi := raw.data.(*btfInt)
-			if bi.Offset() > 0 || bi.Bits().Bytes() != size {
-				legacyBitfields[id] = [2]Bits{bi.Offset(), bi.Bits()}
+			size := header.Size()
+			if err := binary.Read(r, bo, &bInt); err != nil {
+				return nil, fmt.Errorf("can't read btfInt, id: %d: %w", id, err)
 			}
-			typ = &Int{name, raw.Size(), bi.Encoding()}
+			if bInt.Offset() > 0 || bInt.Bits().Bytes() != size {
+				legacyBitfields[id] = [2]Bits{bInt.Offset(), bInt.Bits()}
+			}
+			typ = &Int{name, header.Size(), bInt.Encoding()}
 
 		case kindPointer:
 			ptr := &Pointer{nil}
-			fixup(raw.Type(), &ptr.Target)
+			fixup(header.Type(), &ptr.Target)
 			typ = ptr
 
 		case kindArray:
-			btfArr := raw.data.(*btfArray)
-			arr := &Array{nil, nil, btfArr.Nelems}
-			fixup(btfArr.IndexType, &arr.Index)
-			fixup(btfArr.Type, &arr.Type)
+			if err := binary.Read(r, bo, &bArr); err != nil {
+				return nil, fmt.Errorf("can't read btfArray, id: %d: %w", id, err)
+			}
+
+			arr := &Array{nil, nil, bArr.Nelems}
+			fixup(bArr.IndexType, &arr.Index)
+			fixup(bArr.Type, &arr.Type)
 			typ = arr
 
 		case kindStruct:
-			members, err := convertMembers(raw.data.([]btfMember), raw.Bitfield())
+			vlen := header.Vlen()
+			if len(bMembers) < vlen {
```

Comment:
I think this could just be `slices.Grow`? Same for the other occurrences.

Reply:
Oh, nice! I didn't know about this one yet

Reply:
Hmm, doesn't quite do what we would want, but made a version that does.

Reply:
It would be useful to know why it doesn't do what we want, and also what it is that we want :D

Reply:
There were two main thing. 

First Grow, as the name suggest grows, so it takes a number of elements to increase by not a target size like we have, and passing a negative number is not allowed so it would not have reduced any code, we would have still hard to add an if statement around it.

Secondly, it increases capacity not length. Adding capacity is handy when doing `append` but when we pass the slice to `io.ReadFull` the length is used as an indicator of how much data to read.

So what we want is to increase the length of the array to a given target if we are currently smaller. I now put that logic in the `ensureSize` function which cleans the code up nicely

Reply:
Isn't the answer to that

```
slices.Grow(s[:0], n)[:n]
```

? That also gets rid of the continuous reslicing of buf, bEnums, etc.

Reply:
That does work! I had not though about fixing it that way, very nice!